### ISSUE TITLE
a struct with channel and queue

### ISSUE THREAD

**Issue by kaxap** [jgallagher/amiquip#25]
I wanted to create a constructor which setups and returns the following struct:

```rust
pub struct SomeConsumer<'a> {
    conn: amiquip::Connection,
    chan: amiquip::Channel,
    queue: Option<amiquip::Queue<'a>>,
    consumer: Option<amiquip::Consumer<'a>>,
}
```
But since queue has reference to its Channel (which is the same as `chan` field), I believe it is not possible to create such struct?
Lib's queue struct:

```rust
/// Handle for a declared AMQP queue.
pub struct Queue<'a> {
    channel: &'a Channel,
    name: String,
    message_count: Option<u32>,
    consumer_count: Option<u32>,
}
```

**Comment by kaxap** [comment]
I believe Consumer and Queue have different lifetimes from SomeConsumer?
When I create channel outside of new() and pass it as an argument with 'a lifetime, then everything works. But I want to have it in one function.

**Comment by jgallagher** [comment]
Yes, this is just a general pain point with structs and lifetimes; a struct cannot (easily) contain both an item and a reference to that item.

I haven't tried, but you might be able to accomplish this with [rental](https://crates.io/crates/rental) or [owning_ref](https://crates.io/crates/owning_ref), which allow bundling of an owner and a reference to it together.

**Comment by davidhorac3** [comment]
Did anyone found a doable solution to this? Wanted to give this library a try, but this is a deal breaker I got stuck on.

**Comment by jgallagher** [comment]
The library currently doesn't support this kind of grouping without using something like `rental` or `owning_ref` (linked above), because a struct cannot contain references to its own fields. This is a not-too-uncommon question with Rust in general; e.g., https://stackoverflow.com/questions/32300132/why-cant-i-store-a-value-and-a-reference-to-that-value-in-the-same-struct has a _very_ detailed answer with a description of the problem and possible solutions.

A different design would be to have `Channels` be wrappers around a private type held in an `Rc`, then `Queue` / `Consumer` / etc. could exist without lifetimes (they'd hold that same `Rc`'d inner type). But that isn't how I was using the library, and I wanted compile-time checks that I didn't leave any channels open by holding on to queues or consumers. I'm not aware of a good way to provide both the reference-style API and a "no references wrapped around `Rc`" API simultaneously, though. :(
